src/lem/interpreter.rs

```diff
@@ -1,206 +1,253 @@
+                for (mptr, ptr) in preimg.iter().zip(vec.iter()) {
+                    bind(bindings, mptr.clone(), *ptr)?;
                 }
-                LEMOP::Unhash4(tgts, src) => {
-                    let src_ptr = src.get_ptr(&ptrs)?;
-                    let Some(idx) = src_ptr.get_index4() else {
-                        bail!(
-                            "{} isn't a Tree4 pointer",
-                            src.name()
-                        );
-                    };
-                    let Some((a, b, c, d)) = store.fetch_4_ptrs(idx) else {
-                        bail!("Couldn't fetch {}'s children", src.name())
-                    };
-                    insert_into_ptrs(&mut ptrs, tgts[0].name().clone(), *a)?;
-                    insert_into_ptrs(&mut ptrs, tgts[1].name().clone(), *b)?;
-                    insert_into_ptrs(&mut ptrs, tgts[2].name().clone(), *c)?;
-                    insert_into_ptrs(&mut ptrs, tgts[3].name().clone(), *d)?;
+                preimages.hash3.push(vec.to_vec());
+                Ok(())
+            }
+            LEMOP::Unhash4(preimg, img) => {
+                let img_ptr = img.get_ptr(bindings)?;
+                let Some(idx) = img_ptr.get_index4() else {
+                    bail!("{img} isn't a Tree4 pointer");
+                };
+                let Some((a, b, c, d)) = store.fetch_4_ptrs(idx) else {
+                    bail!("Couldn't fetch {img}'s children")
+                };
+                let vec = [*a, *b, *c, *d];
+                for (mptr, ptr) in preimg.iter().zip(vec.iter()) {
+                    bind(bindings, mptr.clone(), *ptr)?;
                 }
-                LEMOP::Hide(tgt, sec, src) => {
-                    let src_ptr = src.get_ptr(&ptrs)?;
-                    let Ptr::Leaf(Tag::Num, secret) = sec.get_ptr(&ptrs)? else {
-                        bail!("{} is not a numeric pointer", sec.name())
-                    };
-                    let z_ptr = store.hydrate_ptr(&src_ptr)?;
-                    let hash =
-                        store
-                            .poseidon_cache
-                            .hash3(&[secret, z_ptr.tag.to_field(), z_ptr.hash]);
-                    let tgt_ptr = Ptr::comm(hash);
-                    store.comms.insert(FWrap::<F>(hash), (secret, src_ptr));
-                    insert_into_ptrs(&mut ptrs, tgt.name().clone(), tgt_ptr)?;
+                preimages.hash4.push(vec.to_vec());
+                Ok(())
+            }
+            LEMOP::Hide(tgt, sec, src) => {
+                let src_ptr = src.get_ptr(bindings)?;
+                let Ptr::Leaf(Tag::Num, secret) = sec.get_ptr(bindings)? else {
+                    bail!("{sec} is not a numeric pointer")
+                };
+                let z_ptr = store.hash_ptr(src_ptr)?;
+                let hash = store
+                    .poseidon_cache
+                    .hash3(&[*secret, z_ptr.tag.to_field(), z_ptr.hash]);
```

Comment:
Here we use the `hash3` method of the poseidon cache, which holds preimages composed of three field elements.

Flipping back and forth between meanings of `hash3` like this is very confusing. This would be even more confusing if commitments were subject to the slot optimization. I guess the initial LEM doesn't need to do this, but in the general case we will want to be able to deal with arbitrary-sized hash preimages (in terms of field elements) — so we should work toward a structure that allows doing this in a clear and consistent way.

Reply:
A more general solution will be left for a later PR, is that okay?